added QnA Maker Trace onto generateAnswer(). generateAnswer() paramet…

Zerryth opened the pull request:
Fixes #637 

## Description
- generates QnA Maker Trace activity when getAnswers() is called
- deprecate generateAnswers() method to favor getAnswers(), which does not have trace
- can view log in emulator

## Specific Changes

- added `getAnswers()` method, which calls `emitTraceInfo()`
-  `@deprecated` added to `generateAnswer()`
- refactored `qnaMaker.ts` and `qnaMaker.test.js` to use async/await pattern
- added unit tests for updated method `getAnswers()`

Zerryth commented:
aye-aye! will do

cleemullins commented on libraries/botbuilder-ai/src/qnaMaker.ts:
```diff
@@ -193,18 +193,25 @@ export class QnAMaker {
      * @param top (Optional) number of answers to return. Defaults to a value of `1`.
```
These @param comments are not in sync with the changed code below. Please make them match.

cleemullins commented on libraries/botbuilder-ai/src/qnaMaker.ts:
```diff
@@ -193,18 +193,25 @@ export class QnAMaker {
      */
-    public generateAnswer(question: string|undefined, top?: number, scoreThreshold?: number): Promise<QnAMakerResult[]> {
+    public generateAnswer(context: TurnContext, top?: number, scoreThreshold?: number): Promise<QnAMakerResult[]> {
+        const question = context.activity.text;
```
I don't think we can make this change, as it's a compatibility break in code that people are likely using. 

The answer is probably: 
1) Update the existing method as best we can. 
2) Introduce a new method with the new signature and have it work correctly.
3) Put a @depricated decerator on the old method. [https://stackoverflow.com/questions/19412660/how-should-i-mark-a-method-as-obsolete-in-js](https://stackoverflow.com/questions/19412660/how-should-i-mark-a-method-as-obsolete-in-js)

stevengum replied:
D'oh! Yes this would be a breaking change. After talking with @Zerryth, she will instead implement the [QnAMaker.GetAnswersAsync()](https://github.com/Microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.AI.QnA/QnAMaker.cs#L108) from the dotnet SDK  in JS and mark QnAMaker.generateAnswer() as deprecated.

Zerryth replied:
Ok. 

@stevengum found that in the C# there's a method `GetAnswersAsync()` that does not exist in the TS version

I'll make a new method `getAnswers()` in TS, deprecate `generateAnswers()`, and write tests for `getAnswers()` method

It `generateAnswer()` will be left as it originally, since it can't `emitTraceInfo()` without accepting the breaking change `TurnContext` parameter type

cleemullins commented:
@Zerryth LGTM. I've signed off. 

@stevengum Steven?

Zerryth commented:
@stevengum see latest commit for your requested changes:
* Null checked for context.activity in `getAnswers()`. 
* Edited documentation on `getAnswers()` to include mention of Trace added and modified deprecated message in `generateAnswers()`
* simplified return statements in `getAnswers()` and `generateAnswer()`
* added missing period in comment
* restyled `if`/`else` statements to MS BF's convention for TS

Zerryth commented:
Ok, pushed changes to clean up nested if statements in `getAnswers()`.

However now Strict Task Master @stevengum wants me to write tests for the `.sort()` in `getAnswers()` and `generateAnswers()`, so ball still in my court pending those changes :)

stevengum commented:
Right now we're trying to get Microsoft/botbuilder-dotnet#1099 merged in which addresses this issue in Microsoft/botbuilder#5143: *add new GetAnswersAsync method with QnAMakerOptions param*.

@Zerryth, we'll need to get those changes in your PR to avoid any breaking changes. Then your PR will also address #551.

Zerryth commented:
(Sorry @stevengum -- made new PR 'cause silliness involving reimaged computer and not putting new code into clone of this branch)

See fixes in PR https://github.com/Microsoft/botbuilder-js/pull/683
